### ISSUE TITLE
Distinguish the potentially-problematic keypairs in :maps

### ISSUE THREAD

**Issue by leonerd** [neovim/neovim#17225]
### Feature already in Vim?

no

### Feature description

nvim version 0.6.1

Setup: attempt to create two different insert-mode maps:

    :imap <Esc> Escape
    :imap <C-[> Ctrl-OpenBracket

Expected output: they remain distinct;

    :imap
    should print them differently

Actual output: The two are treated as the same, and the latter one wins:

    :imap
    ...
    i  <Esc>  Ctrl-OpenBracket

Two other problematic pairs also collide:

    :imap <Tab> Tab
    :imap <C-I> Ctrl-I

    :imap <CR> CarriageReturn
    :imap <C-M> Ctrl-M

One key pair does seem to behave fine though:

    :imap <BS> Backspace
    :imap <C-H> Ctrl-H

retain their individual identity.

Note this is entirely an issue about the ability to enter and store :maps, and has nothing to do with whether or not a user on an actual terminal can really type them. I am aware that some terminals cannot express the difference between these keys. Some terminals can, as well as most GUI environments; so in those situations users ought to be able to make use of both keys they can type individually

**Comment by zeertzjq** [comment]
Sounds similar to #10079.

**Comment by leonerd** [comment]
@zeertzjq Yes; that looks like a possible partial duplicate. #10079 only refers to one problematic pair (Tab vs Ctrl-I) whereas this refers to all three remaining ones.

**Comment by zeertzjq** [comment]
One problem here is that the translation method used by `:map`s also applies to double quoted string constants in Vim script, and some Vim scripts (including some oldtests) rely on the current behavior.

**Comment by zeertzjq** [comment]
One problem here is that the translation performed by `:map`s also applies to double quoted string constants in Vim script, and some Vim scripts (including some oldtests) rely on the current behavior (`"\<Esc>"` and `"\<C-[>"` both being the same as `"\e"`, `"\<CR>"` and `"\<C-M>"` both being the same as `"\r"`, `"\<Tab>"` and `"\<C-I>"` both being the same as `"\t"`).

**Comment by leonerd** [comment]
I wouldn't mind having to set some option to say "yes I know these keys are really different, I don't mind if those scripts break".

**Comment by clason** [comment]
I think we're all agreed that having these different keys be distinguished is desirable; tests can be changed (although it's tedious), and breaking plugins can and should be adapted.

(That being said, a compatibility option may be helpful for the transition, but probably not at the cost of significant code complexity.)

**Comment by bfredl** [comment]
> Vim scripts (including some oldtests) rely on the current behavior ("\<Esc>" and "\<C-[>" both being the same as "\e", "\<CR>" and "\<C-M>" both being the same as "\r", "\<Tab>" and "\<C-I>" both being the same as "\t").

I'd say we should consider that an anti-pattern for a script to rely on the exact byte representation of `"\<key>"`. if a script wants something guaranteed to be an `\e` byte it should just simply specify `"\e"` and same with `"\r"` and `"\t"` and so on. It's a breaking change in the "spacebar heating" sense only.

**Comment by zeertzjq** [comment]
Not considering any backward compatibility, a solution can be as simple as this:
```diff
diff --git a/src/nvim/getchar.c b/src/nvim/getchar.c
index 3e0f01852..17fc5e61e 100644
--- a/src/nvim/getchar.c
+++ b/src/nvim/getchar.c
@@ -1591,6 +1591,14 @@ int vgetc(void)
         c = utf_ptr2char(buf);
       }
 
+      if ((mod_mask & MOD_MASK_CTRL) && (c >= '?' && c <= '_')) {
+        c = Ctrl_chr(c);
+        mod_mask &= ~MOD_MASK_CTRL;
+        if (c == 0) {  // <C-@> is <Nul>
+          c = K_ZERO;
+        }
+      }
+
       // If mappings are enabled (i.e., not Ctrl-v) and the user directly typed
       // something with a meta- or alt- modifier that was not mapped, interpret
       // <M-x> as <Esc>x rather than as an unbound meta keypress. #8213
diff --git a/src/nvim/keymap.c b/src/nvim/keymap.c
index 32f2158d7..e98d6dc3b 100644
--- a/src/nvim/keymap.c
+++ b/src/nvim/keymap.c
@@ -744,13 +744,8 @@ static int extract_modifiers(int key, int *modp)
       modifiers &= ~MOD_MASK_SHIFT;
     }
   }
-  if ((modifiers & MOD_MASK_CTRL)
-      && ((key >= '?' && key <= '_') || ASCII_ISALPHA(key))) {
-    key = Ctrl_chr(key);
-    modifiers &= ~MOD_MASK_CTRL;
-    if (key == 0) {  // <C-@> is <Nul>
-      key = K_ZERO;
-    }
+  if ((modifiers & MOD_MASK_CTRL) && ASCII_ISALPHA(key)) {
+    key = TOUPPER_ASC(key);
   }
 
   *modp = modifiers;
```

This is probably too breaking though, as it changes the meaning of 26 key notations, not just three.

**Comment by bfredl** [comment]
> This is probably too breaking though, as it changes the meaning of 26 key notations, not just three.

Yes, I think we only change the encoding of the  `<c-x>` that conflict with specific keys, and leave the rest unmodified, even if the internal logic becomes a bit more involved.

**Comment by zeertzjq** [comment]
Oh, I just found [this Vim patch](https://github.com/vim/vim/commit/fccd93f0917234b962ce07d1df3adf9d7105936f). Maybe this notation can be used somewhere.

**Comment by zeertzjq** [comment]
Closed by #17825
